# SQL Analysis using PostgreSQL (Online Retail Cleaned Data)

In this notebook, we will perform **Basic to Advanced Queries** by connecting to the database in **PostgreSQL**. 

---

### 1. Connect to PostgreSQL

We are using **SQLALCHEMY** library (for SQL connection) and **psycopg2-binary** (PostgreSQL adapter for Python).

In [1]:
pip install sqlalchemy psycopg2-binary

Note: you may need to restart the kernel to use updated packages.


In [2]:
from sqlalchemy import create_engine
from sqlalchemy import text
import pandas as pd

Here **create_engine** sets up the database connection and **text (optional here)** is used for executing raw SQL (will be needed for creating **View**.

In [3]:
engine = create_engine('postgresql://postgres:postgres123@localhost:5432/db_retail')

I have created a test database and test connection in PostgreSQL, for doing the end-to-end data fetching by SQL.
This creates a connection engine to a local PostgreSQL database:

**Username: postgres
Password: postgres123
Host: localhost
Port: 5432
Database: db_retail**

In [4]:
#Loads the CSV file into a DataFrame.

df = pd.read_csv('Online_Retail_Clean_Data.csv', dtype={'Invoice': str})
df.columns = (
    df.columns
    .str.strip()                            
    .str.lower()                           
    .str.replace(r'\s+', '_', regex=True)  
    .str.replace(r'[^\w]', '', regex=True) 
)
df.to_sql('online_retail_data', engine, if_exists='replace', index=False)

371

### 2. SQL Queries

In [5]:
pd.read_sql("SELECT * FROM online_retail_data LIMIT 1;", engine).columns

Index(['invoice', 'stockcode', 'description', 'quantity', 'invoicedate',
       'price', 'customer_id', 'country', 'revenue', 'yearmonth',
       'iscancelled'],
      dtype='object')

In [6]:
# Query 1: Top 10 Products by Revenue
pd.read_sql("""
SELECT description, ROUND(SUM(revenue)::numeric, 0) AS total_revenue
FROM online_retail_data
WHERE iscancelled = false
GROUP BY description
ORDER BY total_revenue DESC
LIMIT 10;
""", engine)

description  total_revenue
0            REGENCY CAKESTAND 3 TIER       344563.0
1                              Manual       340731.0
2                      DOTCOM POSTAGE       322657.0
3  WHITE HANGING HEART T-LIGHT HOLDER       266924.0
4         PAPER CRAFT , LITTLE BIRDIE       168470.0
5             JUMBO BAG RED RETROSPOT       150936.0
6                       PARTY BUNTING       149187.0
7       ASSORTED COLOUR BIRD ORNAMENT       132188.0
8                             POSTAGE       127597.0
9     PAPER CHAIN KIT 50'S CHRISTMAS        123142.0

Here, **REGENCY CAKESTAND 3 TIER** is the top product.

In [7]:
# Query 2: Orders per Country
pd.read_sql("""
SELECT country, COUNT(DISTINCT invoice) AS order_count
FROM online_retail_data
GROUP BY country
ORDER BY order_count DESC;
""", engine)

country  order_count
0         United Kingdom        49108
1                Germany         1095
2                   EIRE          806
3                 France          746
4            Netherlands          250
5                  Spain          188
6                Belgium          183
7                 Sweden          129
8               Portugal          124
9            Switzerland          123
10             Australia          117
11                 Italy           92
12       Channel Islands           79
13               Finland           68
14                 Japan           56
15               Denmark           53
16                Norway           53
17               Austria           51
18                Cyprus           45
19                Poland           38
20           Unspecified           28
21                   USA           25
22  United Arab Emirates           22
23             Hong Kong           21
24                Greece           21
25                 Malta           16
26             Singapore           14
27               Bahrain           12
28                Israel           11
29               Iceland            8
30                Canada            8
31             Lithuania            6
32        Czech Republic            5
33    European Community            5
34                   RSA            5
35               Nigeria            3
36               Lebanon            3
37                 Korea            3
38              Thailand            2
39                Brazil            2
40          Saudi Arabia            2
41           West Indies            1
42               Bermuda            1

**United Kingdom** is holding the highest position with order count **49108** and **West Indies, Bermuda** are in the lowest position. 

In [8]:
# Query 3: Total Items Sold per Product
pd.read_sql("""
SELECT stockcode, SUM(quantity) AS total_sold
FROM online_retail_data
GROUP BY stockcode
ORDER BY total_sold DESC;
""", engine)

stockcode  total_sold
0        84077    108545.0
1       85123A     96066.0
2       85099B     95739.0
3        21212     95450.0
4        84879     80705.0
...        ...         ...
5300     10120     -8350.0
5301     23003     -8516.0
5302     23005    -14418.0
5303     22126    -16400.0
5304     20852    -18800.0

[5305 rows x 2 columns]

Here, we can see a total of **5305 unique products (stockcodes)**, each with a corresponding `total_sold` value. Some products have negative totals, which likely represent returns or cancellations.

In [9]:
# Query 4: List of Cancelled Orders
pd.read_sql("""
SELECT *
FROM online_retail_data
WHERE iscancelled = true;
""", engine)

invoice stockcode                       description  quantity  \
0      C489449     22087          PAPER BUNTING WHITE LACE       -12   
1      C489449    85206A      CREAM FELT EASTER EGG BASKET        -6   
2      C489449     21895     POTTING SHED SOW 'N' GROW SET        -4   
3      C489449     21896                POTTING SHED TWINE        -6   
4      C489449     22083        PAPER CHAIN KIT RETRO SPOT       -12   
...        ...       ...                               ...       ...   
19489  C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
19490  C581499         M                            Manual        -1   
19491  C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
19492  C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
19493  C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

               invoicedate   price  customer_id         country  revenue  \
0      2009-12-01 10:33:00    2.95        16321       Australia   -35.40   
1      2009-12-01 10:33:00    1.65        16321       Australia    -9.90   
2      2009-12-01 10:33:00    4.25        16321       Australia   -17.00   
3      2009-12-01 10:33:00    2.10        16321       Australia   -12.60   
4      2009-12-01 10:33:00    2.95        16321       Australia   -35.40   
...                    ...     ...          ...             ...      ...   
19489  2011-12-09 09:57:00    0.83        14397  United Kingdom    -9.13   
19490  2011-12-09 10:28:00  224.69        15498  United Kingdom  -224.69   
19491  2011-12-09 11:57:00   10.95        15311  United Kingdom   -54.75   
19492  2011-12-09 11:58:00    1.25        17315  United Kingdom    -1.25   
19493  2011-12-09 11:58:00    1.25        17315  United Kingdom    -6.25   

      yearmonth  iscancelled  
0       2009-12         True  
1       2009-12         True  
2       2009-12         True  
3       2009-12         True  
4       2009-12         True  
...         ...          ...  
19489   2011-12         True  
19490   2011-12         True  
19491   2011-12         True  
19492   2011-12         True  
19493   2011-12         True  

[19494 rows x 11 columns]

Total **19494** products are cancelled.

In [10]:
# Query 5: Monthly Revenue Trend
pd.read_sql("""
SELECT yearmonth, ROUND(SUM(revenue)::numeric, 0) AS monthly_revenue
FROM online_retail_data
WHERE iscancelled = false
GROUP BY yearmonth
ORDER BY yearmonth;
""", engine)

yearmonth  monthly_revenue
0    2009-12         825686.0
1    2010-01         652709.0
2    2010-02         553340.0
3    2010-03         833570.0
4    2010-04         627935.0
5    2010-05         659859.0
6    2010-06         752270.0
7    2010-07         606681.0
8    2010-08         697275.0
9    2010-09         924333.0
10   2010-10        1126558.0
11   2010-11        1470272.0
12   2010-12        1262599.0
13   2011-01         691365.0
14   2011-02         523632.0
15   2011-03         717639.0
16   2011-04         537809.0
17   2011-05         770536.0
18   2011-06         761740.0
19   2011-07         719221.0
20   2011-08         737014.0
21   2011-09        1058590.0
22   2011-10        1154979.0
23   2011-11        1509496.0
24   2011-12         638811.0

In this dataset, we have data from **Year 2009 to 2011** and this is the monthly revenue trend of total **Two Years**|

In [11]:
# Query 6: Customer Lifetime Value
pd.read_sql("""
SELECT customer_id, ROUND(SUM(revenue)::numeric, 0) AS lifetime_value
FROM online_retail_data
WHERE iscancelled = false AND customer_id !=0
GROUP BY customer_id
ORDER BY lifetime_value DESC
LIMIT 10;
""", engine)

customer_id  lifetime_value
0        18102        608822.0
1        14646        528603.0
2        14156        313946.0
3        14911        295973.0
4        17450        246973.0
5        13694        196483.0
6        17511        175604.0
7        16446        168473.0
8        16684        147143.0
9        12415        144458.0

Here we identified the top 10 customers by **lifetime value**, which represents the total revenue a customer has generated from all non-cancelled orders. Customer lifetime value (CLV) is a key business metric used to understand how valuable a customer is over the duration of their relationship. In this case, customer **18102** is the most valuable, contributing over **608,000** in total revenue.


In [12]:
# Query 7: Revenue by Country (Excluding UK)
pd.read_sql("""
SELECT country,ROUND(SUM(revenue)::numeric, 0) AS total_revenue
FROM online_retail_data
WHERE iscancelled = false AND country != 'United Kingdom'
GROUP BY country
ORDER BY total_revenue DESC;
""", engine)

country  total_revenue
0                   EIRE       664432.0
1            Netherlands       554232.0
2                Germany       431262.0
3                 France       356945.0
4              Australia       169968.0
5                  Spain       109179.0
6            Switzerland       101011.0
7                 Sweden        91904.0
8                Denmark        69862.0
9                Belgium        65753.0
10                Norway        60110.0
11              Portugal        58017.0
12                 Japan        47138.0
13       Channel Islands        44997.0
14                 Italy        32550.0
15               Finland        29926.0
16             Singapore        25317.0
17                Cyprus        24980.0
18             Hong Kong        23894.0
19               Austria        23613.0
20                Greece        19096.0
21                Israel        11335.0
22  United Arab Emirates        11098.0
23           Unspecified        10936.0
24                Poland        10902.0
25                   USA         8367.0
26                 Malta         8099.0
27             Lithuania         6554.0
28               Iceland         5633.0
29                Canada         4883.0
30                   RSA         4804.0
31               Bahrain         3110.0
32              Thailand         3071.0
33               Lebanon         1906.0
34                Brazil         1412.0
35    European Community         1300.0
36               Bermuda         1253.0
37                 Korea         1119.0
38        Czech Republic          827.0
39           West Indies          536.0
40          Saudi Arabia          146.0
41               Nigeria          140.0

We calculated total revenue by country, excluding the UK and cancelled orders. **EIRE, Netherlands, and Germany** have the highest sales outside the UK.


In [13]:
# Query 8: Average Order Value
pd.read_sql("""
SELECT ROUND(AVG(order_total)::numeric, 0) AS avg_order_value
FROM (
  SELECT invoice, SUM(revenue) AS order_total
  FROM online_retail_data
  WHERE iscancelled = false
  GROUP BY invoice
) AS order_summary;
""", engine)

avg_order_value
0            459.0

We calculated the **average order value** across all non-cancelled invoices. The result shows that each order is worth approximately **$459** on average.


In [14]:
# Query 9: Most Frequently Returned Products
pd.read_sql("""
SELECT description, COUNT(*) AS return_count
FROM online_retail_data
WHERE iscancelled = true
GROUP BY description
ORDER BY return_count DESC
LIMIT 10;
""", engine)

description  return_count
0                              Manual           538
1            REGENCY CAKESTAND 3 TIER           347
2                             POSTAGE           229
3       BAKING SET 9 PIECE RETROSPOT            211
4      STRAWBERRY CERAMIC TRINKET BOX           184
5                            Discount           172
6  WHITE HANGING HEART T-LIGHT HOLDER           135
7                 WHITE CHERRY LIGHTS           121
8            RED RETROSPOT CAKE STAND           111
9                             SAMPLES           101

This is the list of top 10 most frequently returned products. **'Manual' and 'REGENCY CAKESTAND 3 TIER'** have the highest return counts.


In [15]:
# Query 10: Segment Orders by High/Low Spend
pd.read_sql("""
SELECT invoice,
       order_value,
       CASE
           WHEN spend_rank = 4 THEN 'Top Spend'
           WHEN spend_rank = 3 THEN 'High'
           WHEN spend_rank = 2 THEN 'Mid'
           ELSE 'Low'
       END AS order_segment
FROM (
    SELECT invoice,
           SUM(revenue) AS order_value,
           NTILE(4) OVER (ORDER BY SUM(revenue)) AS spend_rank
    FROM online_retail_data
    WHERE iscancelled = false
      AND revenue > 0
    GROUP BY invoice
) ranked_orders
WHERE order_value > 0;
""", engine)

invoice  order_value order_segment
0      528127         0.19           Low
1      570554         0.38           Low
2      567869         0.40           Low
3      502731         0.42           Low
4      539441         0.42           Low
...       ...          ...           ...
40072  533027     49844.99     Top Spend
40073  576365     50653.91     Top Spend
40074  574941     52940.94     Top Spend
40075  541431     77183.60     Top Spend
40076  581483    168469.60     Top Spend

[40077 rows x 3 columns]

Here we segmented the orders **Low, Mid, High, and Top Spend using quartiles**. It helps identify high-value transactions.


In [16]:
# Query 11: Most Active Buying Hours
pd.read_sql("""
SELECT EXTRACT(HOUR FROM invoicedate::timestamp) AS hour,
       COUNT(*) AS order_count
FROM online_retail_data
WHERE iscancelled = false
GROUP BY hour
ORDER BY hour;
""", engine)

hour  order_count
0    6.0            1
1    7.0         1064
2    8.0        15689
3    9.0        67325
4   10.0        91221
5   11.0       119213
6   12.0       157226
7   13.0       147823
8   14.0       139399
9   15.0       133406
10  16.0        92298
11  17.0        56144
12  18.0        16449
13  19.0         8571
14  20.0         2048

By the output, the peak order activity occurs between **11:00 AM and 3:00 PM**, with the highest at **12:00 PM**.


In [17]:
# Query 12: Top 3 Customers by Country
pd.read_sql("""
SELECT country, customer_id, total_spent FROM (
  SELECT country, customer_id,
         SUM(revenue) AS total_spent,
         RANK() OVER (PARTITION BY country ORDER BY ROUND(SUM(revenue)::numeric, 0) DESC) AS rnk
  FROM online_retail_data
  WHERE iscancelled = false AND customer_id !=0
  GROUP BY country, customer_id
) AS ranked
WHERE rnk <= 3;
""", engine)

country  customer_id  total_spent
0     Australia        12415    144458.37
1     Australia        12431      8125.54
2     Australia        12388      3901.11
3       Austria        12360      4252.89
4       Austria        12358      3887.07
..          ...          ...          ...
93  Unspecified        12363       552.00
94          USA        12733      2424.87
95          USA        12607      1579.51
96          USA        12549      1575.68
97  West Indies        18140       536.41

[98 rows x 3 columns]

Here is the list of **top 3 spending customers in each country**, ranked by total revenue. It helps identify the most valuable customers per region. We used **RANK() function** for identifying the top customers.


In [18]:
# Query 13: Top 10 Countries by Revenue (Excluding UK)
pd.read_sql("""
SELECT country, ROUND(SUM(Revenue)::numeric, 0) AS TotalRevenue
FROM online_retail_data
WHERE country != 'United Kingdom' AND iscancelled = false
GROUP BY country
ORDER BY TotalRevenue DESC
LIMIT 10;
""", engine)

country  totalrevenue
0         EIRE      664432.0
1  Netherlands      554232.0
2      Germany      431262.0
3       France      356945.0
4    Australia      169968.0
5        Spain      109179.0
6  Switzerland      101011.0
7       Sweden       91904.0
8      Denmark       69862.0
9      Belgium       65753.0

These are top 10 countries by total revenue (excluding the UK). **EIRE, Netherlands, and Germany** are the highest contributors outside the UK.


In [19]:
# Query 14: Repeat vs One-Time Customers
pd.read_sql("""
SELECT
  CASE
    WHEN InvoiceCount = 1 THEN 'One-Time'
    ELSE 'Repeat'
  END AS CustomerType,
  COUNT(*) AS NumCustomers
FROM (
  SELECT customer_id, COUNT(DISTINCT invoice) AS InvoiceCount
  FROM online_retail_data
  WHERE iscancelled = false AND customer_id !=0
  GROUP BY customer_id
) AS sub
GROUP BY CustomerType;
""", engine)

customertype  numcustomers
0     One-Time          1626
1       Repeat          4255

We can see that most customers are repeat buyers, with **4,255 repeat** and **1,626 one-time** customers, indicating strong **customer retention.**

In [20]:
# Query 15: Cancellation Rate by Country
pd.read_sql("""
SELECT country,
       COUNT(*) AS TotalOrders,
       COUNT(CASE WHEN iscancelled THEN 1 END) AS CancelledOrders,
       ROUND(COUNT(CASE WHEN iscancelled THEN 1 END)::numeric * 100 / COUNT(*), 2) AS CancelRate
FROM online_retail_data
GROUP BY Country
ORDER BY CancelRate DESC
LIMIT 10;
""", engine)

country  totalorders  cancelledorders  cancelrate
0                   USA          535              126       23.55
1                 Japan          582               97       16.67
2        Czech Republic           30                5       16.67
3                 Korea           63               10       15.87
4          Saudi Arabia           10                1       10.00
5  United Arab Emirates          500               33        6.60
6               Nigeria           32                2        6.25
7       Channel Islands         1664               95        5.71
8                 Malta          299               17        5.69
9               Germany        17624              921        5.23

We calculated the **cancellation rate by country** using SQL aggregation and conditional logic.
Notably, the **USA** has the highest cancellation rate at **23.55%**, followed by **Japan** and **Czech Republic**, indicating potential regional issues with order fulfillment, payment, or customer satisfaction.


In [21]:
# Query 16: Top 10 spenders using a CTE
pd.read_sql("""
WITH customer_spending AS (
    SELECT 
        customer_id, 
        SUM(revenue) AS total_spent
    FROM online_retail_data
    WHERE iscancelled = false AND customer_id != 0
    GROUP BY customer_id
)
SELECT * 
FROM customer_spending
ORDER BY total_spent DESC
LIMIT 10;
""", engine)

customer_id  total_spent
0        18102    608821.65
1        14646    528602.52
2        14156    313946.37
3        14911    295972.63
4        17450    246973.09
5        13694    196482.81
6        17511    175603.55
7        16446    168472.50
8        16684    147142.77
9        12415    144458.37

Here are the **top 10 highest-spending customers** based on total revenue from completed (non-cancelled) transactions. Using a **Common Table Expression (CTE)**, we group the data by `customer_id` and sum the `revenue` to calculate `total_spent`.